File: app/src/main/java/app/familygem/share/SharingActivity.java
package app.familygem.share;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import org.apache.commons.net.ftp.FTP;
import org.apache.commons.net.ftp.FTPClient;
import org.folg.gedcom.model.Gedcom;
import org.folg.gedcom.model.Header;
import org.folg.gedcom.model.Person;
import org.folg.gedcom.model.Submitter;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

import app.familygem.BaseActivity;
import app.familygem.BuildConfig;
import app.familygem.Exporter;
import app.familygem.Global;
import app.familygem.Principal;
import app.familygem.R;
import app.familygem.Settings;
import app.familygem.U;
import app.familygem.constant.Choice;
import app.familygem.constant.Extra;
import app.familygem.constant.Json;
import app.familygem.list.SubmittersFragment;
import app.familygem.util.ChangeUtils;
import app.familygem.util.TreeUtils;

public class SharingActivity extends BaseActivity {

    Gedcom gc;
    Settings.Tree tree;
    Exporter exporter;
    String submitterName;
    int accessible; // 0 = false, 1 = true
    String dateId;
    String submitterId;
    boolean successfulUpload; // To avoid uploading twice

    @Override
    protected void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.sharing_activity);
        final int treeId = getIntent().getIntExtra(Extra.TREE_ID, 1);
        tree = Global.settings.getTree(treeId);
        final EditText titleEdit = findViewById(R.id.share_title);
        titleEdit.setText(tree.title);

        if (tree.grade == 10)
            ((TextView)findViewById(R.id.share_submitter_title)).setText(R.string.changes_submitter);

        exporter = new Exporter(this);
        exporter.openTree(treeId);
        gc = Global.gc;
        if (gc != null) {
            displayShareRoot();
            // Submitter name
            final Submitter[] submitter = new Submitter[1];
            // albero in Italia con submitter referenziato
            if (tree.grade == 0 && gc.getHeader() != null && gc.getHeader().getSubmitter(gc) != null)
                submitter[0] = gc.getHeader().getSubmitter(gc);
                // in Italia ci sono autori ma nessuno referenziato, prende l'ultimo
            else if (tree.grade == 0 && !gc.getSubmitters().isEmpty())
                submitter[0] = gc.getSubmitters().get(gc.getSubmitters().size() - 1);
                // in Australia ci sono autori freschi, ne prende uno
            else if (tree.grade == 10 && U.autoreFresco(gc) != null)
                submitter[0] = U.autoreFresco(gc);
            final EditText editaAutore = findViewById(R.id.share_submitter);
            submitterName = submitter[0] == null ? "" : submitter[0].getName();
            editaAutore.setText(submitterName);

            // Display an alert for the acknowledgment of sharing
            if (!Global.settings.shareAgreement) {
                new AlertDialog.Builder(this).setTitle(R.string.share_sensitive)
                        .setMessage(R.string.aware_upload_server)
                        .setPositiveButton(android.R.string.ok, (dialog, id) -> {
                            Global.settings.shareAgreement = true;
                            Global.settings.save();
                        }).setNeutralButton(R.string.remind_later, null).show();
            }

            // Raccoglie i dati della condivisione e posta al database
            findViewById(R.id.share_button).setOnClickListener(button -> {
                if (successfulUpload)
                    concludeShare();
                else {
                    if (controlla(titleEdit, R.string.please_title) || controlla(editaAutore, R.string.please_name))
                        return;

                    button.setEnabled(false);
                    findViewById(R.id.share_wheel).setVisibility(View.VISIBLE);

                    // Titolo dell'albero
                    String titoloEditato = titleEdit.getText().toString();
                    if (!tree.title.equals(titoloEditato)) {
                        tree.title = titoloEditato;
                        Global.settings.save();
                    }

                    // Aggiornamento del submitter
                    Header header = gc.getHeader();
                    if (header == null) {
                        header = TreeUtils.INSTANCE.createHeader(tree.id + ".json");
                        gc.setHeader(header);
                    } else
                        header.setDateTime(ChangeUtils.INSTANCE.actualDateTime());
                    if (submitter[0] == null) {
                        submitter[0] = SubmittersFragment.createSubmitter(null);
                    }
                    if (header.getSubmitterRef() == null) {
                        header.setSubmitterRef(submitter[0].getId());
                    }
                    String nomeAutoreEditato = editaAutore.getText().toString();
                    if (!nomeAutoreEditato.equals(submitterName)) {
                        submitterName = nomeAutoreEditato;
                        submitter[0].setName(submitterName);
                        ChangeUtils.INSTANCE.updateChangeDate(submitter[0]);
                    }
                    submitterId = submitter[0].getId();
                    TreeUtils.INSTANCE.saveJsonAsync(gc, treeId); // baypassando la preferenza di non salvare in atomatico

                    // Tree accessibility for app developer
                    CheckBox accessibleTree = findViewById(R.id.share_allow);
                    accessible = accessibleTree.isChecked() ? 1 : 0;

                    // Sends
                    if (!BuildConfig.PASS_KEY.isEmpty()) {
                        new SendingShareData().execute(this);
                    } else {
                        restore(this);
                        Toast.makeText(this, R.string.something_wrong, Toast.LENGTH_LONG).show();
                    }
                }
            });
        } else
            findViewById(R.id.share_layout).setVisibility(View.GONE);
    }

    // The person root of the tree
    View rootView;

    void displayShareRoot() {
        String rootId;
        if (tree.shareRoot != null && gc.getPerson(tree.shareRoot) != null)
            rootId = tree.shareRoot;
        else if (tree.root != null && gc.getPerson(tree.root) != null) {
            rootId = tree.root;
            tree.shareRoot = rootId; // per poter condividere subito l'albero senza cambiare la radice
        } else {
            rootId = U.trovaRadice(gc);
            tree.shareRoot = rootId;
        }
        Person person = gc.getPerson(rootId);
        if (person != null && tree.grade < 10) { // viene mostrata solo alla prima condivisione, non al ritorno
            LinearLayout rootLayout = findViewById(R.id.share_root);
            rootLayout.removeView(rootView);
            rootLayout.setVisibility(View.VISIBLE);
            rootView = U.placePerson(rootLayout, person);
            rootView.setOnClickListener(v -> {
                Intent intent = new Intent(this, Principal.class);
                intent.putExtra(Choice.PERSON, true);
                startActivityForResult(intent, 5007);
            });
        }
    }

    // Verifica che un campo sia compilato
    boolean controlla(EditText campo, int msg) {
        String testo = campo.getText().toString();
        if (testo.isEmpty()) {
            campo.requestFocus();
            InputMethodManager imm = (InputMethodManager)getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.showSoftInput(campo, InputMethodManager.SHOW_IMPLICIT);
            Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
            return true;
        }
        return false;
    }

    /**
     * Inserts the share summary in the database of www.familygem.app.
     * If all goes well creates the ZIP file with the tree and images.
     */
    static class SendingShareData extends AsyncTask<SharingActivity, Void, SharingActivity> {
        @Override
        protected SharingActivity doInBackground(SharingActivity... contesti) {
            SharingActivity activity = contesti[0];
            try {
                String protocol = "https";
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) protocol = "http";
                URL url = new URL(protocol + "://www.familygem.app/insert_share.php");
                HttpURLConnection connection = (HttpURLConnection)url.openConnection();
                connection.setRequestMethod("POST");
                OutputStream stream = new BufferedOutputStream(connection.getOutputStream());
                BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(stream, StandardCharsets.UTF_8));
                String query = "passKey=" + URLEncoder.encode(BuildConfig.PASS_KEY, "UTF-8") +
                        "&treeTitle=" + URLEncoder.encode(activity.tree.title, "UTF-8") +
                        "&submitterName=" + URLEncoder.encode(activity.submitterName, "UTF-8") +
                        "&accessible=" + activity.accessible;
                writer.write(query);
                writer.flush();
                writer.close();
                stream.close();

                // Answer
                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                String line = reader.readLine();
                reader.close();
                connection.disconnect();
                if (line.startsWith("20")) {
                    activity.dateId = line.replaceAll("[-: ]", "");
                    Settings.Share share = new Settings.Share(activity.dateId, activity.submitterId);
                    activity.tree.addShare(share);
                    Global.settings.save();
                } else U.toast(line);
            } catch (Exception e) {
                U.toast(e.getLocalizedMessage());
                e.printStackTrace();
            }
            return activity;
        }

        @Override
        protected void onPostExecute(SharingActivity activity) {
            if (activity.dateId != null && activity.dateId.startsWith("20")) {
                File treeFile = new File(activity.getCacheDir(), activity.dateId + ".zip");
                if (activity.exporter.exportZipBackup(activity.tree.shareRoot, 9, Uri.fromFile(treeFile))) {
                    new FtpUpload().execute(activity);
                    return;
                } else
                    Toast.makeText(activity, activity.exporter.errorMessage, Toast.LENGTH_LONG).show();
            }
            // Un Toast di errore qui sostituirebbe il messaggio di tosta() in catch()
            restore(activity);
        }
    }

    /**
     * Uploads via FTP the ZIP file with the shared tree.
     */
    static class FtpUpload extends AsyncTask<SharingActivity, Void, SharingActivity> {
        protected SharingActivity doInBackground(SharingActivity... activities) {
            SharingActivity activity = activities[0];
            JSONObject credential = U.getCredential(Json.FTP);
            if (credential != null) {
                try {
                    FTPClient ftpClient = new FTPClient();
                    ftpClient.connect(credential.getString(Json.HOST), credential.getInt(Json.PORT));
                    ftpClient.enterLocalPassiveMode();
                    ftpClient.login(credential.getString(Json.USER), credential.getString(Json.PASSWORD));
                    ftpClient.changeWorkingDirectory(credential.getString(Json.SHARED_PATH));
                    ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
                    BufferedInputStream buffIn;
                    String nomeZip = activity.dateId + ".zip";
                    buffIn = new BufferedInputStream(new FileInputStream(activity.getCacheDir() + "/" + nomeZip));
                    activity.successfulUpload = ftpClient.storeFile(nomeZip, buffIn);
                    buffIn.close();
                    ftpClient.logout();
                    ftpClient.disconnect();
                } catch (Exception e) {
                    U.toast(e.getLocalizedMessage());
                }
            }
            return activity;
        }

        protected void onPostExecute(SharingActivity activity) {
            if (activity.successfulUpload) {
                Toast.makeText(activity, R.string.correctly_uploaded, Toast.LENGTH_SHORT).show();
                activity.concludeShare();
            } else {
                restore(activity);
            }
        }
    }

    /**
     * Show apps to share the link.
     */
    private void concludeShare() {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.sharing_tree));
        intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.click_this_link,
                "https://www.familygem.app/share.php?tree=" + dateId));
        //startActivity( Intent.createChooser( intent, "Condividi con" ) );
        /* Tornando indietro da una app di messaggistica il requestCode 35417 arriva sempre corretto
            Invece il resultCode può essere RESULT_OK o RESULT_CANCELED a capocchia
            Ad esempio da Gmail ritorna indietro sempre con RESULT_CANCELED sia che l'email è stata inviata o no
            anche inviando un Sms ritorna RESULT_CANCELED anche se l'sms è stato inviato
            oppure da Whatsapp è RESULT_OK sia che il messaggio è stato inviato o no
            In pratica non c'è modo di sapere se nella app di messaggistica il messaggio è stato inviato */
        startActivityForResult(Intent.createChooser(intent, getText(R.string.share_with)), 35417);
        restore(this);
    }

    private static void restore(SharingActivity activity) {
        activity.findViewById(R.id.share_button).setEnabled(true);
        activity.findViewById(R.id.share_wheel).setVisibility(View.GONE);
    }

    // Aggiorna le preferenze così da mostrare la nuova radice scelta in PersonsFragment
    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == AppCompatActivity.RESULT_OK) {
            if (requestCode == 5007) {
                tree.shareRoot = data.getStringExtra(Extra.RELATIVE_ID);
                Global.settings.save();
                displayShareRoot();
            }
        }
        // Ritorno indietro da qualsiasi app di condivisione, nella quale il messaggio è stato inviato oppure no
        if (requestCode == 35417) {
            // Todo chiudi tastiera
            Toast.makeText(getApplicationContext(), R.string.sharing_completed, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem i) {
        onBackPressed();
        return true;
    }
}

File: app/src/main/java/app/familygem/share/SharingActivity.kt
package app.familygem.share

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import app.familygem.BaseActivity
import app.familygem.Exporter
import app.familygem.Global
import app.familygem.R
import app.familygem.Settings
import app.familygem.Settings.Share
import app.familygem.constant.Choice
import app.familygem.constant.Extra
import app.familygem.databinding.SharingActivityBinding
import app.familygem.main.MainActivity
import app.familygem.main.SubmittersFragment
import app.familygem.util.ChangeUtil.actualDateTime
import app.familygem.util.ChangeUtil.updateChangeDate
import app.familygem.util.PersonUtil
import app.familygem.util.TreeUtil
import app.familygem.util.TreeUtil.createHeader
import app.familygem.util.Util
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okio.buffer
import okio.source
import org.folg.gedcom.model.Gedcom
import org.folg.gedcom.model.Submitter
import org.json.JSONObject
import java.io.File
import java.util.concurrent.TimeUnit
import kotlin.math.ceil
import kotlin.math.min

/** Allows to share a tree by uploading it to the online server. */
class SharingActivity : BaseActivity(R.string.share_tree) {

    private var gedcom: Gedcom? = null
    private lateinit var tree: Settings.Tree
    private lateinit var exporter: Exporter
    private lateinit var titleView: EditText
    private var rootView: View? = null // The person root of the tree
    private lateinit var submitterName: String
    private lateinit var submitterId: String
    private var accessible = false
    private lateinit var apiToken: String
    private var dateId: String? = null
    private var successfulUpload = false // To avoid uploading twice

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent(SharingActivityBinding.inflate(layoutInflater).root)
        val treeId = intent.getIntExtra(Extra.TREE_ID, 1)
        tree = Global.settings.getTree(treeId)
        titleView = findViewById(R.id.share_title)
        titleView.setText(tree.title)
        if (tree.grade == 10) findViewById<TextView>(R.id.share_submitter_title).setText(R.string.changes_submitter)
        exporter = Exporter(this, progressView)
        lifecycleScope.launch(IO) {
            exporter.openTree(treeId)
            gedcom = Global.gc
            withContext(Main) {
                if (gedcom != null) setupInterface()
                else findViewById<View>(R.id.share_layout).visibility = View.GONE
            }
        }
    }

    private fun setupInterface() {
        displayShareRoot()
        // Submitter name
        var submitter: Submitter? = null
        val freshSubmitter = getFreshSubmitter()
        // Tree in Italy with submitter selected as main
        if (tree.grade == 0 && gedcom!!.header != null && gedcom!!.header.getSubmitter(gedcom) != null)
            submitter = gedcom!!.header.getSubmitter(gedcom)
        // In Italy there are submitters but no main one: takes the last one
        else if (tree.grade == 0 && gedcom!!.submitters.isNotEmpty())
            submitter = gedcom!!.submitters[gedcom!!.submitters.size - 1]
        // In Australia there are fresh submitters: takes one
        else if (tree.grade == 10 && freshSubmitter != null)
            submitter = freshSubmitter
        val submitterView = findViewById<EditText>(R.id.share_submitter)
        submitterName = if (submitter == null) "" else submitter.name
        submitterView.setText(submitterName)

        // Displays an alert for the acknowledgment of sharing
        if (!Global.settings.shareAgreement) {
            AlertDialog.Builder(this).setTitle(R.string.share_sensitive)
                .setMessage(R.string.aware_upload_server)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    Global.settings.shareAgreement = true
                    Global.settings.save()
                }.setNeutralButton(R.string.remind_later, null).show()
        }

        // Collects sharing data and posts to database
        findViewById<Button>(R.id.share_button).setOnClickListener { button: View ->
            if (successfulUpload) { // Tree already uploaded: we only share the link
                concludeShare()
            } else {
                if (checkCompiled(titleView, R.string.please_title) || checkCompiled(submitterView, R.string.please_name))
                    return@setOnClickListener

                button.isEnabled = false
                progressView.visibility = View.VISIBLE

                // Tree title
                val editedTitle = titleView.text.trim().toString()
                if (tree.title != editedTitle) {
                    tree.title = editedTitle
                    Global.settings.save()
                }

                // Submitter update
                var header = gedcom!!.header
                if (header == null) {
                    header = createHeader(tree.id.toString() + ".json")
                    gedcom!!.header = header
                } else header.dateTime = actualDateTime()
                if (submitter == null) {
                    submitter = SubmittersFragment.createSubmitter(null)
                }
                if (header.submitterRef == null) {
                    header.submitterRef = submitter!!.id
                }
                val editedSubmitterName = submitterView.text.trim().toString()
                if (editedSubmitterName != submitterName) {
                    submitterName = editedSubmitterName
                    submitter!!.name = submitterName
                    updateChangeDate(submitter)
                }
                submitterId = submitter!!.id
                GlobalScope.launch(IO) {
                    TreeUtil.saveJson(gedcom!!, tree.id) // Possibly bypassing the preference not to save automatically
                }

                // Tree accessibility for app developer
                val accessibleTree = findViewById<CheckBox>(R.id.share_allow)
                accessible = accessibleTree.isChecked

                // Sends
                lifecycleScope.launch(IO) {
                    sendShareData()
                        .onSuccess {
                            successfulUpload = true
                            Util.toast(R.string.correctly_uploaded)
                            restore()
                            concludeShare()
                        }.onFailure {
                            it.printStackTrace()
                            Util.toast(it.message)
                            restore()
                        }
                }
            }
        }
    }

    /** Places a small person card that represents the root of the shared tree. */
    private fun displayShareRoot() {
        val rootId = if (gedcom!!.getPerson(tree.shareRoot) != null) tree.shareRoot
        else if (gedcom!!.getPerson(tree.root) != null) tree.root
        else TreeUtil.findRootId(gedcom!!) // Root is missing or tree has no people
        tree.shareRoot = rootId // In case it was different
        val person = gedcom!!.getPerson(rootId)
        if (person != null && tree.grade < 10) { // Shown only at first sharing, not coming back
            val rootLayout = findViewById<LinearLayout>(R.id.share_root)
            rootLayout.visibility = View.VISIBLE
            rootLayout.removeView(rootView)
            rootView = PersonUtil.placeSmallPerson(rootLayout, person)
            rootView!!.setOnClickListener {
                val intent = Intent(this, MainActivity::class.java)
                intent.putExtra(Choice.PERSON, true)
                startActivityForResult(intent, 5007)
            }
        }
    }

    /** Finds the first non-passed submitter. */
    private fun getFreshSubmitter(): Submitter? {
        for (submitter in gedcom!!.submitters) {
            if (submitter.getExtension("passed") == null) return submitter
        }
        return null
    }

    /** Verifies that a EditText is filled in. */
    private fun checkCompiled(field: EditText, message: Int): Boolean {
        if (field.text.isBlank()) {
            field.requestFocus()
            val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            imm.showSoftInput(field, InputMethodManager.SHOW_IMPLICIT)
            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
            return true
        }
        return false
    }

    /**
     * Inserts the share summary in the database of www.familygem.app.
     * If all goes well creates the ZIP file with the tree and images.
     */
    private suspend fun sendShareData(): Result<Boolean> {
        return try {
            apiToken = Util.getToken()
            val jsonObject = JSONObject().apply {
                put("title", tree.title)
                put("submitterName", submitterName)
                put("accessible", accessible)
            }
            val requestBody = jsonObject.toString().toRequestBody("application/json; charset=utf-8".toMediaTypeOrNull())
            val request = Request.Builder()
                .url(Global.apiUrl + "trees")
                .header("Authorization", "Bearer $apiToken")
                .post(requestBody).build()
            Global.okHttpClient.newCall(request).execute().use { response ->
                val json = JSONObject(response.body.string())
                if (response.isSuccessful) {
                    dateId = json.getString("dateId")
                    val share = Share(dateId, submitterId)
                    tree.addShare(share)
                    Global.settings.save()
                    val treeFile = File(cacheDir, "$dateId.zip")
                    if (exporter.exportZipBackup(tree.shareRoot, 9, Uri.fromFile(treeFile))) {
                        uploadChunkedTree(treeFile)
                        Result.success(true)
                    } else {
                        throw Exception(exporter.errorMessage)
                    }
                } else {
                    throw Exception(json.optString("detail", response.message))
                }
            }
        } catch (exception: Exception) {
            Result.failure(exception)
        }
    }

    /** Uploads the tree ZIP file chunked in parts. */
    private suspend fun uploadChunkedTree(zipFile: File) {
        val totalSize = zipFile.length()
        progressView.displayBar("Uploading tree", totalSize)
        val chunkSize = (2 * 1024 * 1024).toLong() // 2 MB
        val totalChunks = ceil(totalSize.toDouble() / chunkSize).toLong()
        var totalBytesWritten = 0L
        var totalBytesUploaded = 0L
        val maxRetries = 3
        val uploadClient = Global.okHttpClient.newBuilder()
            .connectTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true).build()
        val deferred = (0 until totalChunks).map { i ->
            lifecycleScope.async(IO) {
                val offset = i * chunkSize
                val currentChunkSize = min(chunkSize, totalSize - offset)
                var retryCount = 0
                var success = false
                while (!success) {
                    try {
                        val requestBody = object : RequestBody() {
                            override fun contentType() = "application/octet-stream".toMediaTypeOrNull()
                            override fun contentLength() = currentChunkSize
                            override fun writeTo(sink: okio.BufferedSink) {
                                zipFile.source().buffer().use { source ->
                                    source.skip(offset)
                                    val buffer = ByteArray(8192) // 8 kB segments
                                    var bytesInChunkWritten = 0L
                                    while (bytesInChunkWritten < currentChunkSize) {
                                        val toRead = min(buffer.size.toLong(), currentChunkSize - bytesInChunkWritten).toInt()
                                        val read = source.read(buffer, 0, toRead)
                                        if (read == -1) break
                                        sink.write(buffer, 0, read)
                                        sink.flush()
                                        bytesInChunkWritten += read
                                        totalBytesWritten += read
                                        progressView.progress = totalBytesWritten
                                    }
                                }
                            }
                        }
                        val request = Request.Builder().url(Global.apiUrl + "files?dateId=$dateId&offset=$offset")
                            .header("Authorization", "Bearer $apiToken")
                            .post(requestBody).build()
                        uploadClient.newCall(request).execute().use { response ->
                            if (response.isSuccessful) {
                                totalBytesUploaded += currentChunkSize
                                success = true
                                if (totalBytesUploaded >= totalSize) {
                                    progressView.hideBar()
                                }
                            } else {
                                val jsonResponse = JSONObject(response.body.string())
                                throw Exception(jsonResponse.optString("detail", response.message))
                            }
                        }
                    } catch (exception: Exception) {
                        retryCount++
                        if (retryCount >= maxRetries) {
                            progressView.hideBar()
                            throw exception
                        }
                        Thread.sleep(1000L * retryCount) // Exponential backoff
                    }
                }
            }
        }
        deferred.awaitAll()
    }

    /** Displays available apps to share the link. */
    private fun concludeShare() {
        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/plain"
        intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.sharing_tree))
        intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.click_this_link, "https://www.familygem.app/share.php?tree=$dateId"))
        /* Tornando indietro da una app di messaggistica il requestCode 35417 arriva sempre corretto
            Invece il resultCode può essere RESULT_OK o RESULT_CANCELED a capocchia
            Ad esempio da Gmail ritorna indietro sempre con RESULT_CANCELED sia che l'email è stata inviata o no
            anche inviando un Sms ritorna RESULT_CANCELED anche se l'sms è stato inviato
            oppure da Whatsapp è RESULT_OK sia che il messaggio è stato inviato o no
            In pratica non c'è modo di sapere se nella app di messaggistica il messaggio è stato inviato. */
        startActivityForResult(Intent.createChooser(intent, getText(R.string.share_with)), 35417)
    }

    private suspend fun restore() {
        withContext(Main) {
            findViewById<View>(R.id.share_button).isEnabled = true
            progressView.visibility = View.GONE
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        // Updates preferences to show the new root person chosen in PersonsFragment
        if (resultCode == RESULT_OK && requestCode == 5007) {
            tree.shareRoot = data!!.getStringExtra(Extra.RELATIVE_ID)
            Global.settings.save()
            displayShareRoot()
        }
        // Coming back from any messaging app, where the message was sent or not
        if (requestCode == 35417) {
            // TODO: close keyboard
            Toast.makeText(this, R.string.sharing_completed, Toast.LENGTH_LONG).show()
        }
    }
}
